subxt/src/storage/storage_key.rs

```diff
@@ -0,0 +1,445 @@
+use crate::{
+    error::{Error, MetadataError, StorageAddressError},
+    utils::{Encoded, Static},
+};
+use scale_decode::{visitor::IgnoreVisitor, DecodeAsType};
+use scale_encode::EncodeAsType;
+use scale_info::{PortableRegistry, TypeDef};
+use scale_value::Value;
+use subxt_metadata::{StorageEntryType, StorageHasher};
+
+use derivative::Derivative;
+
+use super::utils::hash_bytes;
+
+#[derive(Debug, Clone)]
+// An iterator over all type ids of the key and the respective hashers.
+pub struct StorageHashersIter {
+    hashers_and_ty_ids: Vec<(StorageHasher, u32)>,
+    idx: usize,
+}
+
+impl StorageHashersIter {
+    /// Creates a new [`StorageHashersIter`]. Looks at the [`StorageEntryType`] and
+    /// assigns a hasher to each type id that makes up the key.
+    pub fn new(storage_entry: &StorageEntryType, types: &PortableRegistry) -> Result<Self, Error> {
+        let mut hashers_and_ty_ids = vec![];
+
+        if let StorageEntryType::Map {
+            hashers, key_ty, ..
+        } = storage_entry
+        {
+            let ty = types
+                .resolve(*key_ty)
+                .ok_or(MetadataError::TypeNotFound(*key_ty))?;
+
+            if let TypeDef::Tuple(tuple) = &ty.type_def {
+                if tuple.fields.len() != hashers.len() {
+                    return Err(StorageAddressError::WrongNumberOfHashers {
+                        hashers: hashers.len(),
+                        fields: tuple.fields.len(),
+                    }
+                    .into());
+                }
+                for (i, f) in tuple.fields.iter().enumerate() {
+                    hashers_and_ty_ids.push((hashers[i], f.id));
+                }
+            } else {
+                if hashers.len() != 1 {
+                    return Err(StorageAddressError::WrongNumberOfHashers {
+                        hashers: hashers.len(),
+                        fields: 1,
+                    }
+                    .into());
+                }
+                hashers_and_ty_ids.push((hashers[0], *key_ty));
+            };
+        }
+
+        Ok(Self {
+            hashers_and_ty_ids,
+            idx: 0,
+        })
+    }
+
+    pub fn next_or_err(&mut self) -> Result<(StorageHasher, u32), Error> {
+        self.next().ok_or_else(|| {
+            StorageAddressError::TooManyKeys {
+                expected: self.hashers_and_ty_ids.len(),
+            }
+            .into()
+        })
+    }
+
+    pub fn reset(&mut self) {
+        self.idx = 0;
+    }
```

Comment:
`reset()` type functions on iterators feel a bit unidiomatic to me (and mutating an iterator scares me a little). Not really a big deal but we could be more idiomatic perhaps by just making the iteration be a separate thing like so:

```rust
// rename `StorageHAshersIter` to this but keep methods the same:
pub struct StorageHashers { 
  // same methods.. 
 
    pub fn iter(&'_ self) -> StorageHashersIter<'_> {
        StorageHashersIter { hashers: self, idx: 0 }
    }
}

// impl `Iterator` on a separate type instead (then we can iterate over them 
// as many times as we want and no iterators are being mutated. Maybe a bit like:
pub struct StorageHashersIter<'a> {
    idx: usize,
    hashers: &'a StorageHashers
}

impl <'a> Iterator for StorageHashersIter<'a> {
    fn next(&mut self) {
        let item = self.hashers.hashers_and_ty_ids.get(self.idx).copied()?;
        self.idx += 1;
        Some(item)
    }
}
```

Reply:
Maybe this is a bit of a pain though, and if so I'm happy to stick with the current approach!

Reply:
Oh yeah, the `reset` is a bit ugly, I implemented a seperate `StorageHashersIter<'a>` struct now.